Add Github Enterprise support

aslafy-z opened the pull request:
Fixes #15

aslafy-z committed:
Add Github Enterprise support

aslafy-z committed:
Update README.md

fregante commented:
I’m thinking that it should just accept a URL and we can parse all the data from it: api, user, repo, directory

Would you be willing to change that?

You can drop the previous user repo and directory parameters and I’ll release a new major version

aslafy-z commented:
Would you expect the user to provide the ref or also infer it from that URL?
I'd suggest to keep the current implementation. It is easier to use for third party use-cases from my point of view.

aslafy-z committed:
Forward api URL

fregante commented:
Yeah, everything this module needs is in the URL, in fact it starts from a URL upstream and that tool has to include a GitHub URL parser and API endpoint selection logic. 

If we really want to preserve that flexibility, the first parameter of this module/function would be `resource`, which can be either a string (the URL) or an object with the 3-5 parts needed. Then it can go back to accepting 2 parameters: resource, options (token)

aslafy-z committed:
Move to unique reference input

aslafy-z commented:
I made the changes, but it still need some validation, and it's untested.

fregante commented on index.js:
```diff
@@ -1,33 +1,59 @@
+			res = {api: 'https://api.github.com'};
+			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname);
+			if (parsedUrl.hostname !== 'github.com') {
+				res.api = `${parsedUrl.protocol}//${parsedUrl.host}/api/v3`;
```
The protocol is always `https`
```suggestion
				res.api = `https://${parsedUrl.host}/api/v3`;
```

fregante replied:
This applies to everywhere you used this, including https://github.com/download-directory/download-directory.github.io/pull/29

fregante commented on index.js:
```diff
@@ -1,33 +1,59 @@
+			const parsedUrl = new URL(res);
+			res = {api: 'https://api.github.com'};
+			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname);
+			if (parsedUrl.hostname !== 'github.com') {
```
I'd rather see a clearer:
```suggestion
			if (parsedUrl.hostname === 'github.com') {
				res.api = 'https://api.github.com';
			} else {
```

fregante commented on index.js:
```diff
@@ -1,33 +1,59 @@
+		try {
+			const parsedUrl = new URL(res);
+			res = {api: 'https://api.github.com'};
+			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname);
```
This way you can avoid the `try`/`catch`

```suggestion
			[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname) || [];
			if (res.directory !== 'string') {
				 throw new Error('Unable to parse GitHub URL'); 
			}
```

fregante commented on index.js:
```diff
@@ -1,33 +1,59 @@
-async function api(endpoint, token) {
-	token = token ? `&access_token=${token}` : '';
-	const response = await fetch(`https://api.github.com/repos/${endpoint}${token}`);
+// Matches '/<re/po>/tree/<ref>/<dir>'
```
```suggestion
// Matches '/<user>/<repo>/tree/<ref>/<dir>'
```

fregante commented on index.js:
```diff
@@ -1,33 +1,59 @@
 	token,
 	getFullData = false
 }) {
+	const res = parseResource(resource);
```
I think you can reuse it:
```suggestion
	resource = parseResource(resource);
```
Perhaps you can actually only parse the string:

```suggestion
	if (typeof resource === 'string') {
		resource = parseResource(resource);
	}
```

fregante commented on index.js:
```diff
@@ -41,21 +67,20 @@ async function viaContentsApi({
+	const res = parseResource(resource);
+
+	if (!res.directory.endsWith('/')) {
+		resource.directory += '/';
```
I think I had made a mistake here, it should be moved after the API call, or else the API endpoint becomes `user/repo//git/trees/ref?recursive=1`, notice the double `/` after repo

aslafy-z replied:
Fixed!

fregante reviewed (changes requested):
Thanks for the changes! You can leave the validation out of this PR for now, it wasn't there before.

aslafy-z committed:
Update index.js

Co-Authored-By: Federico Brigante <github@bfred.it>

aslafy-z committed:
Positive conditional for domain matching

aslafy-z committed:
Drop unecessary comments

aslafy-z committed:
Force https usage

aslafy-z committed:
Apply comments

aslafy-z committed:
Fix missing keyword

aslafy-z committed:
Fix lint

aslafy-z committed:
Fix lint

aslafy-z commented:
Not sure about that lint error:
```
  ✖  11:3  new Error() is too unspecific for a type check. Use new TypeError() instead.  unicorn/prefer-type-error
```

I disabled that line from now.

fregante commented on index.js:
```diff
@@ -1,33 +1,57 @@
+	[, res.user, res.repository, res.ref, res.directory] = urlParserRegex.exec(parsedUrl.pathname) || [];
+	if (typeof res.directory !== 'string') {
+		// eslint-disable-next-line unicorn/prefer-type-error
+		throw new Error('Unable to parse GitHub URL');
```
This fixes the lint

```suggestion
		throw new TypeError('Unable to parse GitHub URL');
```

aslafy-z replied:
I wasn't sure whether it was correct to call it a TypeError. Ok for me.

fregante commented on index.js:
```diff
@@ -1,33 +1,57 @@
-	repository,
-	ref = 'HEAD',
-	directory,
+	resource: res,
```
I think this could just be called `resource`?

```suggestion
	resource,
```

aslafy-z replied:
It makes the references to that object below much more uselessly long. So I preferred to alias it. Let me know if you want me to change it.

fregante replied:
It's `resource` in the first place, we have two names for the same thing. Also this rule will be enabled soon: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/prevent-abbreviations.md

aslafy-z replied:
Fixed!

fregante reviewed (commented):
Other than this it looks good!

Can you also extend this to test URLs as well? (of course not GHE)
https://github.com/fregante/list-github-dir-content/blob/master/tests.js

aslafy-z committed:
Apply lint fix

aslafy-z committed:
Revert "Apply comments"

This reverts commit 8aebf35cc72e19f5c552cdb2d69cfa5960b8cb32.

aslafy-z committed:
Fix tests

aslafy-z committed:
Fix lint

aslafy-z committed:
Fix missed bug

aslafy-z commented:
I had to revert some changes because API url was missing in case an object was passed without API in it, for it to default to Github public API, and ref needed default value too.

aslafy-z committed:
Prevent abrieviations

fregante commented:
I published this as a prerelease. You can plug it into `download-directory` so we can test it live

```sh
npm install list-github-dir-content@3.0.0-0
```

fregante committed:
Merge remote-tracking branch 'origin/master' into pr/16

fregante committed:
Use `resource.origin` instead of `resource.api`

Users shouldn't have to find the exact  GHE API endpoint. We can figure it out from the hostname

fregante committed:
Lint

fregante reviewed (changes requested):
Can you review this? My tests seem to work, but I don't have GHE.

Also the readme needs to be updated, all the properties should be moved under the single `resource` item

aslafy-z commented:
I'd love to see this working. I missed your message 3y ago and was pretty sure the PR was merged. Are you ok to implement this feature today?
